pallets/anonymity-mining/src/lib.rs

```diff
@@ -192,6 +305,171 @@ pub mod pallet {
 }
 
 impl<T: Config<I>, I: 'static> Pallet<T, I> {
+	// Add deposit nullifier hash
+	fn add_deposit_nullifier_hash(
+		id: T::TreeId,
+		nullifier_hash: T::Element,
+	) -> Result<(), DispatchError> {
+		DepositNullifierHashes::<T, I>::insert(id, nullifier_hash, true);
+		Ok(())
+	}
+
+	// Add withdraw nullifier hash
+	fn add_withdraw_nullifier_hash(
+		id: T::TreeId,
+		nullifier_hash: T::Element,
+	) -> Result<(), DispatchError> {
+		WithdrawNullifierHashes::<T, I>::insert(id, nullifier_hash, true);
+		Ok(())
+	}
+
+	/// Store nullifier hashes in deposit and withdraw reward trees
+	fn store_nullifier_hashes(
+		id: T::TreeId,
+		deposit_nullifier_hash: T::Element,
+		withdraw_nullifier_hash: T::Element,
+	) -> Result<(), DispatchError> {
+		Self::add_deposit_nullifier_hash(id, deposit_nullifier_hash)?;
+		Self::add_withdraw_nullifier_hash(id, withdraw_nullifier_hash)?;
+		Ok(())
+	}
+
+	pub fn handle_proof_verification(
+		proof_data: &ProofData<T::Element>,
+	) -> Result<(), DispatchError> {
+		// Stubbed out for now
+		Ok(())
+	}
+
+	/// Update vanchor
+	fn update_vanchor(
+		tree_id: T::TreeId,
+		merkle_root: T::Element,
+		src_resource_id: ResourceId,
+		latest_leaf_index: T::LeafIndex,
+	) -> DispatchResultWithPostInfo {
+		let src_chain_id: T::ChainId = src_resource_id.typed_chain_id().chain_id().into();
+		if T::VAnchor::has_edge(tree_id, src_chain_id) {
+			T::VAnchor::update_edge(
+				tree_id,
+				src_chain_id,
+				merkle_root,
+				latest_leaf_index,
+				src_resource_id,
+			)?;
+			Self::deposit_event(Event::AnchorEdgeUpdated);
+		} else {
+			T::VAnchor::add_edge(
+				tree_id,
+				src_chain_id,
+				merkle_root,
+				latest_leaf_index,
+				src_resource_id,
+			)?;
+			Self::deposit_event(Event::AnchorEdgeAdded);
+		}
+		Ok(().into())
+	}
+
+	fn update_roots(
+		src_tree_id: T::TreeId,
+		dest_tree_id: T::TreeId,
+		src_root_index: T::RootIndex,
+		dest_root_index: T::RootIndex,
+		src_deposit_root: T::Element,
+		src_withdraw_root: T::Element,
+		dest_deposit_root: T::Element,
+		dest_withdraw_root: T::Element,
+	) {
+		// Update roots
```

Comment:
I don't understand why we are updating the deposit / withdraw tree roots twice for a `src` and `dest`. Can you please explain? In my opinion, we should have two functions, one for inserting into the cached deposit root storage and one for inserting into the cached withdraw root storage.